File: Python3/0451_sort_characters_by_frequency.py
# https://leetcode-cn.com/problems/sort-characters-by-frequency/

class Solution:
    def frequencySort(self, s):
        """
        Inputs:
            s: str
        Outputs:
            res: str
        """

        count = {}
        for i in s:
            count[i] = count.get(i,0) + 1

        # Sort according to occurrence frequency.
        count_sort = sorted(count.items(), key=lambda x:-x[1])

        res = ''
        for i in count_sort:
            res = res + i[0] * i[1]

        return res

File: Cpp/0451_sort_characters_by_frequency.cpp
// https://leetcode.cn/problems/sort-characters-by-frequency/

class Solution {
public:
    string frequencySort(string s) {
        unordered_map<char, int> cnt;
        for (char ch: s) {
            cnt[ch] += 1;
        }
        vector<pair<char, int>> cnt_sort;
        for (auto it = cnt.begin(); it != cnt.end(); ++it) {
            cnt_sort.push_back(make_pair(it->first, it->second));
        }
        sort(cnt_sort.begin(), cnt_sort.end(),
            [](const pair<char,int>& p1, const pair<char,int>& p2) {
                return p1.second > p2.second;
            });

        string res;
        for (pair<char,int>& p: cnt_sort) {
            for (int i = 0; i < p.second; ++i) {
                res += p.first;
            }
        }
        return res;
    }
};
